File: app/src/main/java/com/bassamalim/athkar/Keeper.java
package com.bassamalim.athkar;

import android.content.Context;
import android.content.SharedPreferences;
import android.location.Location;

import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.PreferenceManager;

import com.bassamalim.athkar.models.MyLocation;
import com.google.gson.Gson;

public class Keeper extends AppCompatActivity {

    private Gson gson;
    private final SharedPreferences pref;
    private String json;

    public Keeper(Context context) {
        pref = PreferenceManager.getDefaultSharedPreferences(context);
    }

    public Keeper(Context context, Location gLocation) {
        pref = PreferenceManager.getDefaultSharedPreferences(context);
        storeLocation(gLocation);
    }

    public void storeLocation(Location gLocation) {
        MyLocation loc = new MyLocation(gLocation);
        SharedPreferences.Editor editor = pref.edit();
        gson = new Gson();
        json = gson.toJson(loc);
        editor.putString("stored location", json);
        editor.apply();
    }

    public Location retrieveLocation() {
        gson = new Gson();
        json = pref.getString("stored location", "");
        MyLocation myLocation = gson.fromJson(json, MyLocation.class);
        return MyLocation.toLocation(myLocation);
    }

}

File: app/src/main/java/bassamalim/hidaya/helpers/Keeper.kt
package bassamalim.hidaya.helpers

import android.content.SharedPreferences
import android.location.Location
import bassamalim.hidaya.Prefs
import bassamalim.hidaya.models.MyLocation
import bassamalim.hidaya.utils.PrefUtils
import com.google.gson.Gson

class Keeper(
    private var pref: SharedPreferences,
    gLocation: Location? = null
) {

    private var gson = Gson()
    private lateinit var locJson: String

    init {
        if (gLocation != null) storeLocation(gLocation)
    }

    /**
     * Store the location in the shared preferences
     *
     * @param gLocation The Location object that you want to store.
     */
    private fun storeLocation(gLocation: Location?) {
        val loc = MyLocation(gLocation!!)
        locJson = gson.toJson(loc)

        pref.edit()
            .putString("stored location", locJson)
            .apply()
    }

    /**
     * Retrieve the stored location from the shared preferences
     *
     * @return The location object.
     */
    fun retrieveLocation(): Location? {
        locJson = PrefUtils.getString(pref, Prefs.StoredLocation)
        val myLocation = gson.fromJson(locJson, MyLocation::class.java)
        return if (myLocation == null) null
        else MyLocation.toLocation(myLocation)
    }

}